File: backend/app/plans.py
"""Plan catalog + branding policy — mirrors src/lib/plans.ts.

CORE PRINCIPLE: branding is decided by VALIDITY DURATION, not price.
  validity < 30 days  -> AK-LOGIC-AI branding only (no custom logo)
  validity >= 30 days -> custom logo + brand name + custom invoice branding
"""
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

CUSTOM_BRANDING_MIN_DAYS = 30
CARRY_FORWARD_WINDOW_DAYS = 3
VALIDITY_ADDON_PRICE = 50
VALIDITY_ADDON_DAYS = 30


@dataclass(frozen=True)
class PlanDef:
    id: str
    name: str
    price: int
    validity_days: int
    credits: int


FREE_PLAN = PlanDef("free", "Free Plan", 0, 0, 0)

PLANS = [
    PlanDef("trial_20", "₹20 Trial", 20, 1, 10),
    PlanDef("starter_50", "₹50 Starter", 50, 3, 30),
    PlanDef("monthly_199", "₹199 Monthly", 199, 30, 300),
    PlanDef("monthly_299", "₹299 Monthly", 299, 30, 600),
    PlanDef("monthly_399", "₹399 Monthly", 399, 30, 1000),
    PlanDef("monthly_900", "₹900 Monthly", 900, 30, 2500),
]

_BY_ID = {p.id: p for p in [FREE_PLAN, *PLANS]}


def plan_by_id(plan_id: str) -> PlanDef | None:
    return _BY_ID.get(plan_id)


def plan_unlocks_branding(validity_days: int) -> bool:
    """Branding unlocked iff plan validity is 30 days or more."""
    return validity_days >= CUSTOM_BRANDING_MIN_DAYS


def now_utc() -> datetime:
    return datetime.now(timezone.utc)


def is_active(expires_at: datetime | None) -> bool:
    return bool(expires_at) and expires_at > now_utc()


def days_remaining(expires_at: datetime | None) -> int:
    if not expires_at:
        return 0
    delta = expires_at - now_utc()
    return max(0, (delta.days + (1 if delta.seconds else 0)))


def branding_enabled(merchant) -> bool:
    """Custom branding active iff plan is active AND validity >= 30 days."""
    return is_active(merchant.plan_expires_at) and plan_unlocks_branding(merchant.plan_validity_days)


def logo_upload_allowed(merchant) -> bool:
    """Logo upload only for merchants on an active monthly (>=30 day) plan."""
    return branding_enabled(merchant)

File: src/lib/plans.ts
/**
 * Plan catalog + branding policy.
 *
 * CORE PRINCIPLE: branding is decided by VALIDITY DURATION, not price.
 *   validity < 30 days  -> AK-LOGIC AI branding only (no custom logo)
 *   validity >= 30 days -> custom logo + brand name + custom invoice branding
 */

export interface PlanDef {
  id: string;
  name: string;
  price: number;        // ₹
  validityDays: number;
  credits: number;      // PDF credits granted
  tag?: string;         // marketing tag
  popular?: boolean;
  best?: boolean;
}

export const DAY_MS = 86400000;
export const CUSTOM_BRANDING_MIN_DAYS = 30;
/** Renewals within this many days of expiry are "timely" (carry-forward UX hint). */
export const CARRY_FORWARD_WINDOW_DAYS = 3;

export const FREE_PLAN: PlanDef = {
  id: 'free',
  name: 'Free Plan',
  price: 0,
  validityDays: 0,
  credits: 0,
};

/** Recharge plans (short-term <30d are trial/starter; >=30d unlock branding). */
export const PLANS: PlanDef[] = [
  { id: 'trial_20', name: '₹20 Trial', price: 20, validityDays: 1, credits: 10, tag: '1 Day' },
  { id: 'starter_50', name: '₹50 Starter', price: 50, validityDays: 3, credits: 30, tag: '3 Days' },
  { id: 'monthly_199', name: '₹199 Monthly', price: 199, validityDays: 30, credits: 300, tag: '30 Days', popular: true },
  { id: 'monthly_299', name: '₹299 Monthly', price: 299, validityDays: 30, credits: 600, tag: '30 Days' },
  { id: 'monthly_399', name: '₹399 Monthly', price: 399, validityDays: 30, credits: 1000, tag: '30 Days' },
  { id: 'monthly_900', name: '₹900 Monthly', price: 900, validityDays: 30, credits: 2500, tag: '30 Days', best: true },
];

/** ₹50 Validity Extension add-on: +30 days, no new credits. */
export const VALIDITY_ADDON = {
  id: 'addon_validity_50',
  name: '₹50 Validity Extension',
  price: 50,
  extendDays: 30,
};

export function planById(id: string): PlanDef | undefined {
  if (id === FREE_PLAN.id) return FREE_PLAN;
  return PLANS.find((p) => p.id === id);
}

/** Branding unlocked iff the plan validity is 30 days or more. */
export function planUnlocksBranding(validityDays: number): boolean {
  return validityDays >= CUSTOM_BRANDING_MIN_DAYS;
}

export function daysRemaining(expiresAt: number): number {
  if (!expiresAt) return 0;
  return Math.max(0, Math.ceil((expiresAt - Date.now()) / DAY_MS));
}

export function isExpired(expiresAt: number): boolean {
  return !expiresAt || expiresAt < Date.now();
}
